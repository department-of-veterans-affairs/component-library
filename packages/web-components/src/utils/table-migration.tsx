/** @jsx React.createElement */
import React from 'react';

/**
 * This is based off of the React `<Table>` implementation where fields
 * was an array of objects w/ `label`, `value`, and a few other properties.
 * `label` is used for display, and value is used to match object properties
 * for each row.
 */
interface FieldType {
  label: string;
  value: string;
}

/**
 * Useful for generating child content for a `<va-table>`
 *
 * @param {array} data - An array of arrays, where the inner arrays represent row data
 * @param {array} fields - The field/column names
 */
export function generateTableChildren(
  data: object[],
  fields: FieldType[],
): JSX.Element[] {
  return [
    /** @ts-ignore */
    <va-table-row slot="headers">
      {fields.map(col => (
        <span>{col.label}</span>
      ))}
      {/** @ts-ignore */}
    </va-table-row>,

    ...data.map(row => {
      return (
        /** @ts-ignore */
        <va-table-row>
          {fields.map(field => (
            <span>{row[field.value]}</span>
          ))}
          {/** @ts-ignore */}
        </va-table-row>
      );
    }),
  ];
}
