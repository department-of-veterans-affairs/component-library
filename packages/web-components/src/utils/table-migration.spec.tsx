/** @jsx React.createElement */
import React from 'react';
import { generateTableChildren } from './table-migration';

describe('generateTableChildren', () => {
  const fields = [
    { label: 'ID', value: 'id' },
    { label: 'Name', value: 'name' },
    { label: 'Role', value: 'role' },
  ];
  const data = [
    { id: 1, name: 'Michael', role: 'Management' },
    { id: 10, name: 'Toby', role: 'HR' },
    { id: 25, name: 'Jim', role: 'Sales' },
  ];

  it('generates <va-table-row> React children>', () => {
    const children = generateTableChildren(data, fields);

    expect(children[0]).toEqual(
      // @ts-ignore
      <va-table-row slot="headers">
        <span>ID</span>
        <span>Name</span>
        <span>Role</span>
        {/* @ts-ignore */}
      </va-table-row>,
    );

    expect(children[1]).toMatchSnapshot();
    expect(children[2]).toMatchSnapshot();
    expect(children[3]).toMatchSnapshot();
  });
});
