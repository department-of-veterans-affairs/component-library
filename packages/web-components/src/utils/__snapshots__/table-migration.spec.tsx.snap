// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTableChildren generates <va-table-row> React children> 1`] = `
<va-table-row>
  <span>
    1
  </span>
  <span>
    Michael
  </span>
  <span>
    Management
  </span>
</va-table-row>
`;

exports[`generateTableChildren generates <va-table-row> React children> 2`] = `
<va-table-row>
  <span>
    10
  </span>
  <span>
    Toby
  </span>
  <span>
    HR
  </span>
</va-table-row>
`;

exports[`generateTableChildren generates <va-table-row> React children> 3`] = `
<va-table-row>
  <span>
    25
  </span>
  <span>
    Jim
  </span>
  <span>
    Sales
  </span>
</va-table-row>
`;
